### PR TITLE
Upcasting

### PR DESCRIPTION
**Pull request by prolic**
resolves https://github.com/prooph/event-store/issues/254

### CODE REVIEW

**Comment by prolic** [comment]
For discussion:

- We need an instance of a message to create a new message - this means when updating events, you need to create a new class, f.e. `UserRegistered` => `UserRegisterd` + `UserRegisteredV2` - is that okay?
- Should we maybe prefer to do upcasting with message name / payload / metadata only? If should remove the message factory from the event store, add our own (maybe `GenericMessageFactory`) - and use the `real` MessageFactory only within this plugin - how can this be achieved, if this is preferred?
- Do we need support for looking into the future of the stream per upcasting? Or can the upcaster itself store some information internally is required?

**Comment by mrook** [comment]
That was fast! I'll take an in-depth look as soon as I have time.

As for the discussion points:

- the new event class needs to exist before the upcaster can work.
- the question is whether you want each upcaster individually to handle deserializing the raw event (so that would indeed mean only passing metadata & payload), or present the deserialized domain event to the upcaster; I think I prefer the first version, as serialization may have changed over time as well.
- some upcasters could need this to merge events, they may be more interested in the past than the future however. In general I think this is something best handled in individual upcasters.

**Comment by prolic** [comment]
@codeliner @mrook tests now included

**Comment by jkuchar** [comment]
Why not put upcast and downcast methods directly into events? These up/downcast methods are tightly coupled with event in particuar version anyway...

**Comment by prolic** [comment]
@jkuchar Single Responsibility Principle

**Comment by mrook** [comment]
@jkuchar because some upcasters would need information from other events (i.e., if you are merging) or emit multiple events (if you are splitting).

File: src/Upcasting/UpcasterChain.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+     * @var Upcaster[]
+     */
+    private $upcasters;
+
+    public function __construct(Upcaster ...$upcasters)
+    {
+        $this->upcasters = $upcasters;
+    }
+
+    public function upcast(Message $message): array
```
Comment (codeliner) [inline src/Upcasting/UpcasterChain.php]: Can you take a second look on the message?

Why is the message turned into a messages array just to foreach over it?

You wanted to return `messages`, right? But `$result` is returned.
Foreach can be removed in either case.

Comment (prolic) [inline src/Upcasting/UpcasterChain.php]: You're right. I will fix this later.

File: src/Upcasting/UpcastingIterator.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public function current(): ?Message
+    {
+        if (! empty($this->storedMessages)) {
+            return reset($this->storedMessages);
+        }
+
+        $messages = $this->upcaster->upcast($this->innerIterator->current());
+
+        if (empty($messages)) {
+            $this->next();
```
Comment (prolic) [inline src/Upcasting/UpcastingIterator.php]: $this->innerIterator->next();

File: src/Upcasting/UpcastingIterator.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+            if ($this->valid()) {
```
Comment (prolic) [inline src/Upcasting/UpcastingIterator.php]: $this->innerIterator->valid()

File: src/Upcasting/UpcasterChain.php (hunk 1 above)
Comment (prolic) [inline src/Upcasting/UpcasterChain.php]: Sorry, you are `not` right :)

At the first iteration of upcasters, you would have been right, but in the second iteration, I might have an array of messages, that may need further upcasting. see: UpcasterChainTest::it_chains_upcasts

This test would immediately break, if we would make the suggested change.

**Comment by prolic** [comment]
There is still a bug in the iterator... I am trying to fix this now.

**Comment by prolic** [comment]
@codeliner I fixed the upcasting iterator

**Comment by prolic** [comment]
@mrook @codeliner your thoughts in the implementation?

**Comment by mrook** [comment]
@prolic looks fine for the initial implementation!

File: src/Upcasting/UpcasterChain.php (hunk 1 above)
Comment (codeliner) [inline src/Upcasting/UpcasterChain.php]: ok, I see. Just a few lines of code but very tricky ;)

**Comment by codeliner** [comment]
LGTM

**Comment by mrook** [comment]
:+1:

**Pull request merged**